src/foam/nanos/auth/UserAndGroupAuthService.js

```diff
@@ -0,0 +1,444 @@
+        // of looking it up in a DAO because if the user is actually an entity that
+        // an agent is acting as, then the user we get from the DAO won't have the
+        // correct group, which is the group set on the junction between the agent
+        // and the entity.
+        User user = (User) x.get("user");
+
+        // check if user exists and is enabled
+        if ( user == null || ! user.getEnabled() ) return false;
+
+        try {
+          String groupId = (String) user.getGroup();
+
+          while ( ! SafetyUtil.isEmpty(groupId) ) {
+            Group group = (Group) groupDAO_.find(groupId);
+
+            // if group is null break
+            if ( group == null ) break;
+
+            // check if group is enabled
+            if ( ! group.getEnabled() ) return false;
+
+            // check permission
+            if ( group.implies(x, permission) ) return true;
+
+            // check parent group
+            groupId = group.getParent();
+          }
+        } catch (IllegalArgumentException e) {
+          Logger logger = (Logger) x.get("logger");
+          logger.error("check", permission, e);
+        } catch (Throwable t) {
+        }
+
+        return false;
+      `
+    },
+    {
+      name: 'check',
+      javaCode: `
+        return checkPermission(x, new AuthPermission(permission));
+      `
+    },
+    {
+      name: 'validatePassword',
+      javaCode: `
+        if ( SafetyUtil.isEmpty(potentialPassword) || ! (Pattern.compile(PASSWORD_VALIDATE_REGEX)).matcher(potentialPassword).matches() ) {
+          throw new RuntimeException(PASSWORD_VALIDATION_ERROR_MESSAGE);
+        }
+      `
+    },
+    {
+      name: 'checkUser',
+      javaCode: `
+        return checkUserPermission(x, user, new AuthPermission(permission));
+      `
+    },
+    {
+      name: 'updatePassword',
+      documentation: `Given a context with a user, validate the password to be
+        updated and return a context with the updated user information.`,
+      javaCode: `
+        if ( x == null || SafetyUtil.isEmpty(oldPassword) || SafetyUtil.isEmpty(newPassword) ) {
+          throw new RuntimeException("Password fields cannot be blank");
+        }
+
+        Session session = x.get(Session.class);
+        if ( session == null || session.getUserId() == 0 ) {
+          throw new AuthenticationException("User not found");
+        }
+
+        User user = (User) userDAO_.find(session.getUserId());
+        if ( user == null ) {
+          throw new AuthenticationException("User not found");
+        }
+
+        // check if user enabled
+        if ( ! user.getEnabled() ) {
+          throw new AuthenticationException("User disabled");
+        }
+
+        // check if user login enabled
+        if ( ! user.getLoginEnabled() ) {
+          throw new AuthenticationException("Login disabled");
+        }
+
+        // check if user group enabled
+        Group group = (Group) groupDAO_.find(user.getGroup());
```

Comment:
```suggestion
        Group group = user.findGroup();
```

Reply:
To avoid CLA issues, I'm not adding your suggestion directly. Thanks!